spring-data-jpa/src/main/java/org/springframework/data/jpa/repository/query/QueryParser.java

```diff
@@ -0,0 +1,159 @@
+import org.springframework.util.Assert;
+
+/**
+ * Operations needed to parse a JPA query.
+ *
+ * @author Greg Turnquist
+ * @since 3.1
+ */
+abstract class QueryParser {
+
+	private final DeclaredQuery declaredQuery;
+
+	QueryParser(DeclaredQuery declaredQuery) {
+		this.declaredQuery = declaredQuery;
+	}
+
+	QueryParser(String query) {
+		this(DeclaredQuery.of(query, false));
+	}
+
+	DeclaredQuery getDeclaredQuery() {
+		return declaredQuery;
+	}
+
+	String getQuery() {
+		return getDeclaredQuery().getQueryString();
+	}
+
+	/**
+	 * Parse the JPA query using its corresponding ANTLR parser.
+	 */
+	abstract ParserRuleContext parse();
+
+	/**
+	 * Generate a query using the original query with an @literal order by} clause added (or amended) based upon the
+	 * provider {@link Sort} parameter.
+	 *
+	 * @param parsedQuery
+	 * @param sort can be {@literal null}
+	 */
+	String createQuery(ParserRuleContext parsedQuery, Sort sort) {
+
+		Assert.notNull(parsedQuery, "parsedQuery cannot be null!");
+		return render(doCreateQuery(parsedQuery, sort));
+	}
+
+	/**
+	 * Generate a count-based query using the original query.
+	 *
+	 * @param parsedQuery
+	 * @param countProjection
+	 */
+	String createCountQuery(ParserRuleContext parsedQuery, @Nullable String countProjection) {
+
+		Assert.notNull(parsedQuery, "parsedQuery cannot be null!");
+		return render(doCreateCountQuery(parsedQuery, countProjection));
+	}
+
+	/**
+	 * Find the projection of the query.
+	 * 
+	 * @param parsedQuery
+	 */
+	String projection(ParserRuleContext parsedQuery) {
+
+		Assert.notNull(parsedQuery, "parsedQuery cannot be null!");
+		return render(doFindProjection(parsedQuery));
+	}
+
+	/**
+	 * Create a {@link QueryParsingToken}-based query with an {@literal order by} applied/amended based upon the
+	 * {@link Sort} parameter.
+	 * 
+	 * @param parsedQuery
+	 * @param sort can be {@literal null}
+	 */
+	abstract List<QueryParsingToken> doCreateQuery(ParserRuleContext parsedQuery, Sort sort);
+
+	/**
+	 * Create a {@link QueryParsingToken}-based count query.
+	 *
+	 * @param parsedQuery
+	 * @param countProjection
+	 */
+	abstract List<QueryParsingToken> doCreateCountQuery(ParserRuleContext parsedQuery, @Nullable String countProjection);
+
+	/**
+	 * Find the alias of the query's primary FROM clause
+	 *
+	 * @return can be {@literal null}
+	 */
+	abstract String findAlias(ParserRuleContext parsedQuery);
+
+	/**
+	 * Find the projection of the query's primary SELECT clause.
+	 *
+	 * @param parsedQuery
+	 */
+	abstract List<QueryParsingToken> doFindProjection(ParserRuleContext parsedQuery);
+
+	/**
+	 * Discern if the query has a {@code new com.example.Dto()} DTO constructor in the select clause.
+	 * 
+	 * @param parsedQuery
+	 * @return Guaranteed to be {@literal true} or {@literal false}.
+	 */
+	abstract boolean hasConstructor(ParserRuleContext parsedQuery);
+
+	private static final Pattern PUNCTUATION_PATTERN = Pattern.compile(".*((?![._])[\\p{Punct}|\\s])");
```

Comment:
Nit: These are at a strange position, should be on top of the class.

Reply:
I can move this toward the top.